src/geometry/manifold/algebra/left_invariant_derivation.lean

```diff
@@ -0,0 +1,192 @@
+/-
+Copyright © 2020 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nicolò Cavalleri
+-/
+
+import geometry.manifold.derivation_bundle
+
+/-!
+
+# Left invariant derivations
+
+In this file we define the concept of left invariant derivation for a Lie group. The concept is
+analogous to the more classical concept of left invariant vector fields, and it holds that the
+derivation associated to a vector field is left invariant iff the the field is.
+
+Moreover we prove that `left_invariant_derivation I G` has the structure of a Lie algebra, hence
+implementing one of the possible definitions of Lie algebra.
+
+-/
+
+noncomputable theory
+
+open_locale lie_group manifold derivation
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]
+{H : Type*} [topological_space H] (I : model_with_corners 𝕜 E H)
+(G : Type*) [topological_space G] [charted_space H G] [monoid G] [has_smooth_mul I G] (g h : G)
+
+-- Generate trivial has_sizeof instance. It prevents weird type class inference timeout problems
+local attribute [nolint instance_priority, instance, priority 10000]
+private def disable_has_sizeof {α} : has_sizeof α := ⟨λ _, 0⟩
+
+/--
+Left-invariant global derivations.
+
+A global derivation is left-invariant if it is equal to its pullback along left multiplication by
+an arbitrary element of `G`.
+-/
+structure left_invariant_derivation extends derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯ :=
+(left_invariant'' : ∀ f : C^∞⟮I, G; 𝕜⟯, ∀ g, 𝒅(𝑳 I g) 1 (derivation.eval_at 1 to_derivation) f =
```

Comment:
I think it might be worth breaking `𝒅(𝑳 I g) 1` as the map:
```lean
point_derivation I (1 : G) →ₗ[𝕜] point_derivation I g
```
I note that since `𝑳 I g 1 = g` though true `by simp` is not true by `rfl` so there might be some trickiness to this but I think it's worth doing.

Reply:
What does breaking out mean?

Reply:
Sorry, I mean making a separate definition for this map.

Reply:
Come to think of it, we should actually have a one-parameter family of equivs:
```
point_derivation I (1 : G) ≃ₗ[𝕜] point_derivation I g
```

This could be out of scope but it might even make sense to define these via the two-parameter families:
```
-- right
point_derivation I g ≃ₗ[𝕜] point_derivation I (g*h)
-- left
point_derivation I g ≃ₗ[𝕜] point_derivation I (h*g)
```
and prove that these have the obvious composition properties.

Reply:
But `𝒅(𝑳 I g) 1` is already a map `point_derivation I (1 : G) →ₗ[𝕜] point_derivation I g`! I mean I defined the differential to give a linear map. Could you make me an example with code of what this separate definition should look like?

Reply:
I think `point_derivation I (1 : G) ≃ₗ[𝕜] point_derivation I g` is worth a separate definition because it is a mathematically important fact in its own right (essentially the parallelisability of tangent bundle of a Lie group) and is being used implicitly here.

In addition, I think it is valuable because it simplifies this definition. I claim that for someone not familiar with this part of the library, the syntax `𝒅(𝑳 I g) 1` is completely opaque. If it is replaced with a separate definition, I think this will help the reader see what is going on more easily.

Reply:
How would you call this map? I feel here there is content for a new PR: to do this in full I need to define a new version of the differential that assigns a linear equiv to a diffeomorphism (an refactor left and right multiplication to be diffeos) and write API for it, and then from this derive the two maps you are talking about and write API for it. I will make a new PR when I have some time! Should it get merged first?

Reply:
> How would you call this map?

Perhaps something like `lie_group.left_mul_point_derivation_equiv`.

> I feel here there is content for a new PR: to do this in full I need to define a new version of the differential that assigns a linear equiv to a diffeomorphism (an refactor left and right multiplication to be diffeos) and write API for it, and then from this derive the two maps you are talking about and write API for it.

You may well be right if all this is required then I agree, but perhaps there is an easier route.

> I will make a new PR when I have some time! Should it get merged first?

This is not my call, but I think it depends on what the alternative options look like.

Reply:
It seems to be that we might quite often be in the position where we know the image of some point under a differentiable function so I wonder if it's worth considering changing the signature of `fdifferential` to:
```lean
def fdifferential (f : C^∞⟮I, M; I', M'⟯) {x : M} {y : M'} (hxy : f x = y) :
  point_derivation I x →ₗ[𝕜] point_derivation I' y :=
{ to_fun := fdifferential_map' f hxy,
  map_smul' := λ k v, rfl,
  map_add' := λ v w, rfl }
```

If we had that and also introduced:
```lean
lemma lie_group.left_mul_id_eq : 𝑳 I g 1 = g := by simp only [L_apply, mul_one]
```
then we have `𝒅(𝑳 I g) (lie_group.left_mul_id_eq I g) : point_derivation I 1 →ₗ[𝕜] point_derivation I g` which would be very convenient.

Reply:
Yes this makes sense (I thought the second lemma you wrote was in mathlib)! At this point we could also take out f as an explicit parameter no? As in `{f}`?

Reply:
Do you still feel like there would be need to define the map you were talking about before if I implemented the differential this way?

Reply:
> I thought the second lemma you wrote was in mathlib

I couldn't find it but I certainly could have missed it.

> At this point we could also take out f as an explicit parameter no? As in `{f : C^∞⟮I, M; I', M'⟯}`?

Possibly. I'm not actually sure what should be implicit / explicit since we should also consider the use case where someone does not know what the image of a point is. In this case the the tautological `rfl : f x = f x` will have to be passed so it might actually make sense to have only `y` implicit to make this convenient.

> Do you still feel like there would be need to define the map you were talking about before if I implemented the differential this way?

Probably not!

Reply:
Done! I added a version of the differential as you requested! I think it has some problems (for example the proposition `f x = y` is shown as a `_`, which makes it hard to carry out a proof containing it) but I agree the advantages outweigh this problem.

Btw note that in this theory the map `𝒅(𝑳 I g) 1` is not necessarily part of an equivalence `point_derivation I (1 : G) ≃ₗ[𝕜] point_derivation I g` because the theory is carried out for a smooth monoid and not for a Lie group. As a counterexample consider the closed interval [0, 1] and take `g = 0`.